suggest to use associated function syntax diagnostic take receiver as first arg mistakely

### Code

```Rust
struct A {}

impl A {
    fn hello(a: i32) {}
}

fn main() {
    let a = A {};
    a.hello(1);
}
```


### Current output

```Shell
error[E0599]: no method named `hello` found for struct `A` in the current scope
  --> src/main.rs:46:7
   |
38 | struct A {}
   | -------- method `hello` not found for this struct
...
46 |     a.hello(1);
   |     --^^^^^---
   |     | |
   |     | this is an associated function, not a method
   |     help: use associated function syntax instead: `A::hello(a, 1)`
   |
   = note: found the following associated functions; to be used as methods, functions must have a `self` parameter
note: the candidate is defined in an impl for the type `A`
  --> src/main.rs:41:5
   |
41 |     fn hello(a: i32) {}
   |     ^^^^^^^^^^^^^^^^
```


### Desired output

```Shell
receiver shouldn't be treated as the first arg of the assoc function when the first arg of the assoc function isn't the same as the current impl type

46 |     a.hello(1);
   |     --^^^^^---
   |     | |
   |     | this is an associated function, not a method
   |     help: use associated function syntax instead: `A::hello(1)`
```


### Rationale and extra context

rustc --version
rustc 1.75.0-nightly (9d83ac217 2023-10-31)

### Other cases

```Rust
no
```


### Anything else?

no

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"Young-Flash"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_hir_typeck/src/method/suggest.rs
+++ b/compiler/rustc_hir_typeck/src/method/suggest.rs
@@ -1591,10 +1591,12 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             {
                 let sig = self.tcx.fn_sig(assoc.def_id).instantiate_identity();
                 sig.inputs().skip_binder().get(0).and_then(|first| {
-                    if first.peel_refs() == rcvr_ty.peel_refs() {
-                        None
-                    } else {
+                    let impl_ty = self.tcx.type_of(*impl_did).instantiate_identity();
+                    // if the type of first arg is the same as the current impl type, we should take the first arg into assoc function
+                    if first.peel_refs() == impl_ty {
                         Some(first.ref_mutability().map_or("", |mutbl| mutbl.ref_prefix_str()))
+                    } else {
+                        None
                     }
                 })
             } else {

--- a/tests/ui/suggestions/suggest-assoc-fn-call-without-receiver.fixed
+++ b/tests/ui/suggestions/suggest-assoc-fn-call-without-receiver.fixed
@@ -1,0 +1,16 @@
+// run-rustfix
+
+struct A {}
+
+impl A {
+    fn hello(_a: i32) {}
+    fn test(_a: Self, _b: i32) {}
+}
+
+fn main() {
+    let _a = A {};
+    A::hello(1);
+    //~^ ERROR no method named `hello` found
+    A::test(_a, 1);
+    //~^ ERROR no method named `test` found
+}

--- a/tests/ui/suggestions/suggest-assoc-fn-call-without-receiver.rs
+++ b/tests/ui/suggestions/suggest-assoc-fn-call-without-receiver.rs
@@ -1,0 +1,16 @@
+// run-rustfix
+
+struct A {}
+
+impl A {
+    fn hello(_a: i32) {}
+    fn test(_a: Self, _b: i32) {}
+}
+
+fn main() {
+    let _a = A {};
+    _a.hello(1);
+    //~^ ERROR no method named `hello` found
+    _a.test(1);
+    //~^ ERROR no method named `test` found
+}

--- a/tests/ui/suggestions/suggest-assoc-fn-call-without-receiver.stderr
+++ b/tests/ui/suggestions/suggest-assoc-fn-call-without-receiver.stderr
@@ -1,0 +1,41 @@
+error[E0599]: no method named `hello` found for struct `A` in the current scope
+  --> $DIR/suggest-assoc-fn-call-without-receiver.rs:12:8
+   |
+LL | struct A {}
+   | -------- method `hello` not found for this struct
+...
+LL |     _a.hello(1);
+   |     ---^^^^^---
+   |     |  |
+   |     |  this is an associated function, not a method
+   |     help: use associated function syntax instead: `A::hello(1)`
+   |
+   = note: found the following associated functions; to be used as methods, functions must have a `self` parameter
+note: the candidate is defined in an impl for the type `A`
+  --> $DIR/suggest-assoc-fn-call-without-receiver.rs:6:5
+   |
+LL |     fn hello(_a: i32) {}
+   |     ^^^^^^^^^^^^^^^^^
+
+error[E0599]: no method named `test` found for struct `A` in the current scope
+  --> $DIR/suggest-assoc-fn-call-without-receiver.rs:14:8
+   |
+LL | struct A {}
+   | -------- method `test` not found for this struct
+...
+LL |     _a.test(1);
+   |     ---^^^^---
+   |     |  |
+   |     |  this is an associated function, not a method
+   |     help: use associated function syntax instead: `A::test(_a, 1)`
+   |
+   = note: found the following associated functions; to be used as methods, functions must have a `self` parameter
+note: the candidate is defined in an impl for the type `A`
+  --> $DIR/suggest-assoc-fn-call-without-receiver.rs:7:5
+   |
+LL |     fn test(_a: Self, _b: i32) {}
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0599`.

